src/router.rs

```diff
@@ -195,13 +197,29 @@ impl<'a, Data> Resource<'a, Data> {
             let new_resource = ResourceData {
                 endpoints: HashMap::new(),
                 middleware: self.middleware_base.clone(),
+                method_options: Vec::new(),
             };
             *resource = Some(new_resource);
         }
         let resource = resource.as_mut().unwrap();
+        if method != http::Method::OPTIONS {
+            // It's important that this panics here, otherwise we will get weird results for OPTIONS
+            // routes
+            if resource.endpoints.contains_key(&method) {
+                panic!("A {} endpoint already exists for this path", method)
+            }
 
-        if resource.endpoints.contains_key(&method) {
-            panic!("A {} endpoint already exists for this path", method)
+            resource.method_options.push(method.as_str().to_string());
+            let option_header = resource.method_options.join(", ");
+            self.options(async || {
```

Comment:
Repeating myself on Discord: try `async move`. I suppose that the error is about lifetime bounds. Since `Endpoint` requires the closure to be `'static`, you should move `option_header` into the closure.

Reply:
Accidentally resolved the conversation, my bad!

I tried that initially. Still got the same error.

Reply:
We might need some trait bounds for `Data` then! I'll try it.

Reply:
I think so too, if you look at [my reply above](https://github.com/rust-net-web/tide/pull/104#issuecomment-444010806) I got the error that `Data` is missing the bounds `Send + Sync + Clone`, but I wasn't sure how I could add bounds to it since `Endpoint` is automatically implemented by the macros in `endpoint.rs`

Reply:
Okay, I *almost* resolved the problem, but the biggest one remains. Here's my diff:

```diff
diff --git a/src/router.rs b/src/router.rs
index 0468a71..f9d609f 100644
--- a/src/router.rs
+++ b/src/router.rs
@@ -173,7 +173,7 @@ struct ResourceData<Data> {
     method_options: Vec<String>,
 }

-impl<'a, Data> Resource<'a, Data> {
+impl<'a, Data: 'static + Clone + Send + Sync> Resource<'a, Data> {
     /// "Nest" a subrouter to the path.
     ///
     /// This method will build a fresh `Router` and give a mutable reference to it to the builder
@@ -211,11 +211,11 @@ impl<'a, Data> Resource<'a, Data> {

             resource.method_options.push(method.as_str().to_string());
             let option_header = resource.method_options.join(", ");
-            self.options(async || {
+            self.options(async move || {
                 http::Response::builder()
                     .status(http::status::StatusCode::OK)
-                    .header("Access-Control-Allow-Methods", option_header)
-                    .header("Allow", option_header)
+                    .header("Access-Control-Allow-Methods", option_header.as_str())
+                    .header("Allow", option_header.as_str())
                     .header("Content-Type", "text/plain")
                     .body(Body::empty())
                     .unwrap()
```

When I build this, I get an awkward error that I can't reason:

```
error[E0525]: expected a closure that implements the `Fn` trait, but this closure only implements `FnOnce`
   --> src/router.rs:214:26
    |
214 |               self.options(async move || {
    |  __________________-------_^^^^^^^^^^^^^_-
    | |                  |       |
    | |                  |       this closure implements `FnOnce`, not `Fn`
    | |                  the requirement to implement `Fn` derives from here
215 | |                 http::Response::builder()
216 | |                     .status(http::status::StatusCode::OK)
217 | |                     .header("Access-Control-Allow-Methods", option_header.as_str())
...   |
221 | |                     .unwrap()
222 | |             });
    | |_____________- closure is `FnOnce` because it moves the variable `option_header` out of its environment
```

As far as I see, there isn't any point that move `option_header` out of the closure. `.as_str()` takes `&self`, and that's it. This is very confusing for me. Maybe `async` is doing something?

Reply:
Oh yeah I completely missed the impl block bounds.

Also, as you mentioned below, why not just remove the CORS header handling from here and introduce it in middleware?

Reply:
Oh wow, thank you guys for figuring that out! I'll remove the CORS specific header, that makes sense but we do still need the `Allow` header. Will that work if we put that back in?

Reply:
This thread seems blocked on a reply from @tirr-c or @bIgBV -- would either of you mind replying to this when you have time?